File: PriorityQueue.py
class arr:
	def __init__(self,data,priority):
		self.data = data
		self.priority = priority
		self.index = -1

class priorityQueue:
	def __init__(self):
		self.Q = list()
		self.heapSize = -1

	def heapify(self,i):
		small = i
		left = 2*i+1
		right = 2*i+2
		
		if self.heapSize >= left and self.Q[left].priority < self.Q[small].priority:
			small = left
			
		if self.heapSize >= right and self.Q[right].priority < self.Q[small].priority:
			small = right
	
		if small != i:
			self.swap(small,i)
			self.heapify(small)

	def swap(self,i,j):
		self.Q[i].data ,self.Q[j].data = self.Q[j].data ,self.Q[i].data
		self.Q[i].priority ,self.Q[j].priority = self.Q[j].priority ,self.Q[i].priority
		self.Q[i].index ,self.Q[j].index = self.Q[j].index ,self.Q[i].index

	def parent(self,i):
		return (i-1)//2

	def enQueue(self,data,priority):
		self.heapSize += 1
		index = self.heapSize

		
		self.Q.append(arr(data,priority))
		self.Q[index].index = index
		p = self.parent(index)
		while  p >= 0:
			if  self.Q[p].priority > self.Q[index].priority:
				self.swap(p,index)
				self.Q[index].index = index 
			index = p

			p = self.parent(index)


	def printHeap(self):
		for i in range(self.heapSize+1):
			# print(self.Q[i].data,end = " ")
			print(str(self.Q[i].data)+"=>"+str(self.Q[i].priority),end = " ")
		print(" ")

	def deQueue(self):
		if self.isEmpty():
			return -1
		temp = self.Q[0]
		self.swap(self.heapSize,0)
		self.heapSize -= 1
		self.heapify(0)
		
		return temp

	def isEmpty(self):
		if self.heapSize == -1:
			return True
		return False

	def updatePriority(self,data,priority):
		for i in self.Q:
			if i.data == data:
				i.priority = priority


if __name__ == "__main__":

	pq =priorityQueue()
	pq.enQueue(1,10)
	print(pq.heapSize)
	pq.printHeap()
	pq.deQueue()
	pq.printHeap()
	print(pq.heapSize)
	pq.enQueue(2,3)
	pq.enQueue(5,6)
	pq.enQueue(3,9)
	pq.printHeap()
	a = pq.deQueue()
	print(a.data,a.priority)
	pq.printHeap()
	a = pq.deQueue()
	# print(a.data,a.priority)
	pq.printHeap()

File: PriorityQueue.c
#include<stdio.h>
#include<stdlib.h>
#include<math.h>

struct heap {
	int* array;
	int heap_size;
	int capacity;
};

struct heap* createHeap(int capacity) {
	struct heap *heap = malloc(sizeof(struct heap));
	if(!heap) {
		printf("memory error");
		return 0;
	}
	heap->array = malloc(sizeof(int)*capacity);
	heap->heap_size = -1;
	heap->capacity = capacity;
	return heap;
}

void resize(struct heap *h) {
	int *old_array = h->array;
	h->capacity = h->capacity*2;
	h->array = malloc(sizeof(int)*h->capacity);
	if(!h->array) {
		printf("memory error");
		return;
	}
	for(int i=0;i<=h->heap_size;i++) 
		h->array[i] = old_array[i];
	printf("\nResizing array size");
	free(old_array);
}

void swap(struct heap *h,int i1,int i2) {
	int temp;
	temp = h->array[i1];
	h->array[i1] = h->array[i2];
	h->array[i2] = temp;
}

void heapify(struct heap *h,int i) {
	int largest = i;
	int left = 2*i+1;
	int right = 2*i+2;

	if (left <= h->heap_size && h->array[left] > h->array[largest] ) 
		largest = left;

	if (right <= h->heap_size && h->array[right] > h->array[largest]) 
		largest = right;

	if(largest!=i) {
		int temp;
		temp = h->array[largest];
		h->array[largest] = h->array[i];
		h->array[i] = temp;
		heapify(h,largest);
	}
}

void build_max_heap(struct heap *h) {
	for(int i = floor(h->heap_size-1)/2;i >= 0;i--) {
		heapify(h,i);
	}
}

int parentIndex(struct heap *h,int i) {
	return (i-1)/2;
}
void insert(struct heap *h,int data) {
	if(h->heap_size == h->capacity-1) 
		resize(h);
	h->array[++h->heap_size] = data;
}

int extractMaxHeap(struct heap *h) {
	if (h->heap_size<0) {
		printf("heap is empty");
		return -1;
	}
	int max = h->array[0];
	swap(h,h->heap_size,0);
	h->heap_size--;
	heapify(h,0);
	return max;
}

int maxOfHeap(struct heap *h) {
	return h->array[0];
}

void increaseHeapData(struct heap *h,int data,int i) {
	if(data < h->array[i]) {
		printf("Given value is small");
		return ;
	}
	h->array[i] = data;
	while(i > 1 && h->array[parentIndex(h,i)] < h->array[i]) {
		swap(h,parentIndex(h,i),i);
		i = parentIndex(h,i);
	}
}

void maxHeapInsert(struct heap *h,int data) {
	h->array[++h->heap_size] = -1;
	increaseHeapData(h,data,h->heap_size);
}

void print(struct heap *h) {
	for(int i=0;i <= h->heap_size;i++) 
		printf("%d ",h->array[i]);
}

void main() {
	struct heap *h = createHeap(6);
	insert(h,6);
	insert(h,5);
	insert(h,4);
	insert(h,3);
	insert(h,2);
	insert(h,1);
	insert(h,0);

	printf("\nheap Elements : ");
	print(h);

	printf("\nmax : %d",maxOfHeap(h));
	printf("\nheap Elements : ");
	print(h);

	printf("\nextract max : %d",extractMaxHeap(h));
	printf("\nheap Elements : ");
	print(h);

	increaseHeapData(h,8,5);
	printf("\nheap Elements : ");
	print(h);

	maxHeapInsert(h,10);
	printf("\nheap Elements : ");
	print(h);

}